### PR TITLE
Two factor authentication - Trust this device

### PR DESCRIPTION
**Pull request by eshiol**
Pull Request for Issue #20150.

### Summary of Changes
This plugin allows user to mark the computer or mobile device as trusted. With trusted computers and devices, you don’t need to enter a verification code for two factor authentication plug-in each time you sign in.

### Testing Instructions
Enable the two factor authentication system
Authenticate yourself with the option trust this device checked
Logout
Authenticate yourself without using the secret key

### Expected result
You can login

### CODE REVIEW

File: administrator/components/com_admin/sql/updates/mysql/3.8.0-2017-09-21.sql [hunk 1]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE `#__content` ADD COLUMN `featured_up` datetime NOT NULL DEFAULT '0000-00-00 00:00:00';
```
Comment (laoneo) [inline administrator/components/com_admin/sql/updates/mysql/3.8.0-2017-09-21.sql]: Should all the featured up and down code be part of this pr?

**Comment by alikon** [comment]
i'm afraid you are mixing things  

```sql 
ALTER TABLE #__content ADD COLUMN featured_up datetime NOT NULL DEFAULT '0000-00-00 00:00:00';
```

File: administrator/components/com_admin/sql/updates/mysql/3.8.0-2017-09-21.sql (hunk 1 above)
Comment (eshiol) [reply administrator/components/com_admin/sql/updates/mysql/3.8.0-2017-09-21.sql]: no

**Comment by eshiol** [comment]
I think so

**Comment by eshiol** [comment]
Sorry guys, my mistake, I fixed the PR

File: administrator/modules/mod_login/tmpl/default.php [hunk 2]
```diff
@@ -63,6 +63,11 @@
 						</label>
 					</span>
 					<input name="secretkey" autocomplete="off" tabindex="3" id="mod-login-secretkey" type="text" class="input-medium" placeholder="<?php echo JText::_('JGLOBAL_SECRETKEY'); ?>" size="15"/>
+					<?php if (JPluginHelper::isEnabled('twofactorauth', 'trust')) : ?>
```
Comment (brianteeman) [inline administrator/modules/mod_login/tmpl/default.php]: not a fan of introducing inline css

File: administrator/modules/mod_login/tmpl/default.php [hunk 3]
```diff
@@ -63,6 +63,11 @@
...
+					<span class="hasTooltip" title="<?php echo JText::_('MOD_LOGIN_TRUST_THIS_DEVICE'); ?>" style="position:absolute;right:70px">
```
Comment (brianteeman) [inline administrator/modules/mod_login/tmpl/default.php]: Not a fan of introducing inline css

File: installation/sql/mysql/joomla.sql [hunk 4]
```diff
@@ -1644,7 +1645,8 @@ INSERT INTO `#__postinstall_messages` (`extension_id`, `title_key`, `description
...
-(700, 'PLG_PLG_RECAPTCHA_VERSION_1_POSTINSTALL_TITLE', 'PLG_PLG_RECAPTCHA_VERSION_1_POSTINSTALL_BODY', 'PLG_PLG_RECAPTCHA_VERSION_1_POSTINSTALL_ACTION', 'plg_captcha_recaptcha', 1, 'action', 'site://plugins/captcha/recaptcha/postinstall/actions.php', 'recaptcha_postinstall_action', 'site://plugins/captcha/recaptcha/postinstall/actions.php', 'recaptcha_postinstall_condition', '3.8.6', 1);
+(700, 'PLG_PLG_RECAPTCHA_VERSION_1_POSTINSTALL_TITLE', 'PLG_PLG_RECAPTCHA_VERSION_1_POSTINSTALL_BODY', 'PLG_PLG_RECAPTCHA_VERSION_1_POSTINSTALL_ACTION', 'plg_captcha_recaptcha', 1, 'action', 'site://plugins/captcha/recaptcha/postinstall/actions.php', 'recaptcha_postinstall_action', 'site://plugins/captcha/recaptcha/postinstall/actions.php', 'recaptcha_postinstall_condition', '3.8.6', 1),
+(700, 'PLG_USER_PRIVACYCONSENT_POSTINSTALL_TITLE', 'PLG_USER_PRIVACYCONSENT_POSTINSTALL_BODY', '', 'plg_user_privacyconsent', 1, 'message', '', '', '', '', '3.9.0', 1);
```
Comment (brianteeman) [inline installation/sql/mysql/joomla.sql]: this line should not be part of this PR

Comment (brianteeman) [reply installation/sql/mysql/joomla.sql]: I think you meant to include the PLG_TWOFACTORAUTH_TRUST_POSTINSTALL_TITLE message and not this one

File: installation/sql/postgresql/joomla.sql [hunk 5]
```diff
@@ -657,6 +657,7 @@ INSERT INTO "#__extensions" ("extension_id", "package_id", "name", "type", "elem
 (478, 0, 'plg_editors-xtd_fields', 'plugin', 'fields', 'editors-xtd', 0, 1, 1, 0, '', '', '', '', 0, '1970-01-01 00:00:00', 0, 0),
 (479, 0, 'plg_sampledata_blog', 'plugin', 'blog', 'sampledata', 0, 1, 1, 0, '', '', '', '', 0, '1970-01-01 00:00:00', 0, 0),
 (480, 0, 'plg_system_sessiongc', 'plugin', 'sessiongc', 'system', 0, 1, 1, 0, '', '', '', '', 0, '1970-01-01 00:00:00', 0, 0),
+(481, 0, 'plg_twofactorauth_trust', 'plugin', 'privacyconsent', 'trust', 0, 0, 1, 0, '', '', '', '', 0, '1970-01-01 00:00:00', 0, 0),
```
Comment (brianteeman) [inline installation/sql/postgresql/joomla.sql]: should be twofactorauth and not privacyconsent

File: installation/sql/sqlazure/joomla.sql [hunk 6]
```diff
@@ -872,6 +872,7 @@ INSERT INTO "#__extensions" ("extension_id", "package_id", "name", "type", "elem
 (478, 0, 'plg_editors-xtd_fields', 'plugin', 'fields', 'editors-xtd', 0, 1, 1, 0, '', '', '', '', 0, '1900-01-01 00:00:00', 0, 0),
 (479, 0, 'plg_sampledata_blog', 'plugin', 'blog', 'sampledata', 0, 1, 1, 0, '', '', '', '', 0, '1900-01-01 00:00:00', 0, 0),
 (480, 0, 'plg_system_sessiongc', 'plugin', 'sessiongc', 'system', 0, 1, 1, 0, '', '', '', '', 0, '1900-01-01 00:00:00', 0, 0),
+(481, 0, 'plg_twofactorauth_trust', 'plugin', 'privacyconsent', 'trust', 0, 0, 1, 0, '', '', '', '', 0, '1970-01-01 00:00:00', 0, 0),
```
Comment (brianteeman) [inline installation/sql/sqlazure/joomla.sql]: should be twofactorauth and not privacyconsent

File: modules/mod_login/tmpl/default.php [hunk 7]
```diff
@@ -82,6 +82,11 @@
 
 			</div>
 		</div>
+			<?php if (JPluginHelper::isEnabled('twofactorauth', 'trust')) : ?>
+		<div id="form-login-trust" class="control-group checkbox">
+			<label for="modlgn-trust" class="control-label"><?php echo JText::_('MOD_LOGIN_TRUST_THIS_DEVICE'); ?></label> <input id="modlgn-trust" type="checkbox" name="trust" class="inputbox" value="yes"/>
```
Comment (brianteeman) [inline modules/mod_login/tmpl/default.php]: I prefer real words for the label and id eg mod-login-trust
it is more consistent with our current uses

File: installation/sql/mysql/joomla.sql (hunk 4 above)
Comment (eshiol) [reply installation/sql/mysql/joomla.sql]: ops, copy and paste is never a good idea

File: plugins/twofactorauth/trust/trust.php [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+	/**
+	 * Database object
+	 *
+	 * @var    JDatabaseDriver
+	 * @since  __DEPLOY_VERSION__
+	 */
+	protected $db;
+
+	/**
+	 * Set the trust this device cookie only after login is successfullly finished.
```
Comment (brianteeman) [inline plugins/twofactorauth/trust/trust.php]: typo 
successfullly ==> successfully

File: modules/mod_login/tmpl/default.php (hunk 7 above)
Comment (eshiol) [reply modules/mod_login/tmpl/default.php]: I used the same style of remember me. Are you sure mod-login-trust is better than modlng-trust?

File: administrator/components/com_admin/sql/updates/mysql/3.9.0-2018-04-18.sql [hunk 9]
```diff
@@ -0,0 +1,5 @@
+INSERT INTO `#__extensions` (`extension_id`, `package_id`, `name`, `type`, `element`, `folder`, `client_id`, `enabled`, `access`, `protected`, `manifest_cache`, `params`, `custom_data`, `system_data`, `checked_out`, `checked_out_time`, `ordering`, `state`) VALUES
+(481, 0, 'plg_twofactorauth_trust', 'plugin', 'trust', 'twofactorauth', 0, 0, 1, 0, '', '', '', '', 0, '0000-00-00 00:00:00', 0, 0);
```
Comment (Quy) [inline administrator/components/com_admin/sql/updates/mysql/3.9.0-2018-04-18.sql]: The upcoming privacy plugin is also 481.

File: plugins/twofactorauth/trust/trust.php [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+				{
+					JLog::add('Invalid cookie detected.', JLog::WARNING, 'error');
+				}
+				else
+				{
+					// Filter series since we're going to use it in the query
+					$filter = new JFilterInput;
+					$old_series = $filter->clean($cookieArray[1], 'ALNUM');
+				}
+				// Destroy the cookie in the browser anyway
```
Comment (Quy) [inline plugins/twofactorauth/trust/trust.php:73]: Add blank line above.

File: plugins/twofactorauth/trust/trust.php [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+		catch (RuntimeException $e)
+		{
+			return false;
+		}
+
+		if (count($results) !== 1)
+		{
+			// Destroy the cookie in the browser.
+			$this->app->input->cookie->set($cookieName, '', 1, $this->app->get('cookie_path', '/'), $this->app->get('cookie_domain', ''));
+			return false;
```
Comment (Quy) [inline plugins/twofactorauth/trust/trust.php:237]: Add blank line above.

File: plugins/twofactorauth/trust/trust.php [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+					$errorCount++;
+
+					// We'll let this query fail up to 5 times before giving up, there's probably a bigger issue at this point
+					if ($errorCount === 5)
+					{
+						return false;
+					}
+				}
+			}
+
```
Comment (Quy) [inline plugins/twofactorauth/trust/trust.php]: Remove blank line.

File: installation/sql/sqlazure/joomla.sql [hunk 13]
```diff
@@ -2342,7 +2343,8 @@ INSERT INTO "#__postinstall_messages" ("extension_id", "title_key", "description
...
 (700, 'TPL_HATHOR_MESSAGE_POSTINSTALL_TITLE', 'TPL_HATHOR_MESSAGE_POSTINSTALL_BODY', 'TPL_HATHOR_MESSAGE_POSTINSTALL_ACTION', 'tpl_hathor', 1, 'action', 'admin://templates/hathor/postinstall/hathormessage.php', 'hathormessage_postinstall_action', 'admin://templates/hathor/postinstall/hathormessage.php', 'hathormessage_postinstall_condition', '3.7.0', 1),
-(700, 'PLG_PLG_RECAPTCHA_VERSION_1_POSTINSTALL_TITLE', 'PLG_PLG_RECAPTCHA_VERSION_1_POSTINSTALL_BODY', 'PLG_PLG_RECAPTCHA_VERSION_1_POSTINSTALL_ACTION', 'plg_captcha_recaptcha', 1, 'action', 'site://plugins/captcha/recaptcha/postinstall/actions.php', 'recaptcha_postinstall_action', 'site://plugins/captcha/recaptcha/postinstall/actions.php', 'recaptcha_postinstall_condition', '3.8.6', 1);
+(700, 'PLG_PLG_RECAPTCHA_VERSION_1_POSTINSTALL_TITLE', 'PLG_PLG_RECAPTCHA_VERSION_1_POSTINSTALL_BODY', 'PLG_PLG_RECAPTCHA_VERSION_1_POSTINSTALL_ACTION', 'plg_captcha_recaptcha', 1, 'action', 'site://plugins/captcha/recaptcha/postinstall/actions.php', 'recaptcha_postinstall_action', 'site://plugins/captcha/recaptcha/postinstall/actions.php', 'recaptcha_postinstall_condition', '3.8.6', 1)
```
Comment (Quy) [inline installation/sql/sqlazure/joomla.sql]: Add comma at the end.

File: administrator/components/com_admin/sql/updates/mysql/3.9.0-2018-04-18.sql (hunk 9 above)
Comment (brianteeman) [reply administrator/components/com_admin/sql/updates/mysql/3.9.0-2018-04-18.sql]: not until it is accepted. :)

This is fine as it is for now

File: modules/mod_login/tmpl/default.php (hunk 7 above)
Comment (brianteeman) [reply modules/mod_login/tmpl/default.php]: it would appear that we are wonderfully inconsistent and use mod-login on the admin  modlgn- on the front end. Oh how I love inconsistency ;)

Comment (dgrammatiko) [reply modules/mod_login/tmpl/default.php]: wasn't that the reason that @ciar4n came up with a solid strategy (the BEM issue)?

File: administrator/modules/mod_login/tmpl/default.php (hunk 3 above)
Comment (zero-24) [reply administrator/modules/mod_login/tmpl/default.php]: Please don't use inline css and move it to a stylesheet. Inline css should not be hardcoded in new places.

**Comment by Quy** [comment]
Tooltip string not translated on backend.

trust

File: modules/mod_login/tmpl/default.php [hunk 14]
```diff
@@ -82,6 +82,11 @@
 
 			</div>
 		</div>
+			<?php if (JPluginHelper::isEnabled('twofactorauth', 'trust')) : ?>
```
Comment (Quy) [inline modules/mod_login/tmpl/default.php:84]: Remove 1 indent to match line 91. Do the same for line 89.

Comment (eshiol) [reply modules/mod_login/tmpl/default.php:84]: it is an if inside another if, why do you want to remove the indentation? What do you think if I add 1 indent to the lines 86, 87 and 88?

Comment (mbabker) [reply modules/mod_login/tmpl/default.php:84]: 86-88 should be indented, as well as 60-84 (unrelated to this PR but those lines are definitely mis-aligned).

Comment (eshiol) [reply modules/mod_login/tmpl/default.php:84]: what about line 82, can I remove that?

**Comment by mbabker** [comment]
Conflicts need resolving.

It'd be nice to get this included...

**Review by brianteeman — APPROVED** [review]
language changes are fine

**Comment by fabpacheco** [comment]
Can you update the sql files extension_id, please. After that I can test this.
Since you created this PR some new extensions were included. The 481 id is assigned to the plg_fields_repeatable in Joomla 3.9 dev.

**Comment by fabpacheco** [comment]
@eshiol thanks for the update but I think you forgot to update the installation sql files too.

I used patchtester and tried to test it but the extension is not in the database.

**Comment by eshiol** [comment]
I have updated the extension_id in the installation sql files but it seems the patch tester component doesn't run the update sql files.
You need to manually run the file 3.9.0-2018-04-18.sql.

**Comment by mbabker** [comment]
> it seems the patch tester component doesn't run the update sql files

That has always been the case.  Patch tester will never arbitrarily run SQL statements, especially as we do not provide a rollback capability.

**Comment by fabpacheco** [comment]
Thanks for the info @mbabker and sorry I'm new to testing here so I did not know that.

**Comment by fabpacheco** [comment]
Ok, I ran the sql in the database and was able to correctly install the extension. I'll do some tests and report the results when finished.

**Comment by fabpacheco** [comment]
Sorry for the delay, this is the tests I made.
Since the pactchtester can't run SQL statements, thanks again @mbabker for the info I tried to use git to install the patch.

**Installation**
Please correct me if I did anything wrong here since this is the first time I installed 3.9-dev using git bash

I used the following commands to install the 3.9 branch
git clone -b 3.9-dev --single-branch https\:\/\/github.com/joomla/joomla-cms.git

Then the following to get this pull request 
git fetch origin pull/20191/head:pr-20191
git checkout pr-20191

I think you probably can install the extension via patchtester and then run the sql but I choose to try the git way

Run the SQL query in your database. In my case In phpMyAdmin operations I ran the following query from @eshiol pr

```
INSERT INTO `#__extensions` (`extension_id`, `package_id`, `name`, `type`, `element`, `folder`, `client_id`, `enabled`, `access`, `protected`, `manifest_cache`, `params`, `custom_data`, `system_data`, `checked_out`, `checked_out_time`, `ordering`, `state`) VALUES
(495, 0, 'plg_twofactorauth_trust', 'plugin', 'trust', 'twofactorauth', 0, 0, 1, 0, '', '', '', '', 0, '0000-00-00 00:00:00', 0, 0);
INSERT INTO `#__postinstall_messages` (`extension_id`, `title_key`, `description_key`, `action_key`, `language_extension`, `language_client_id`, `type`, `action_file`, `action`, `condition_file`, `condition_method`, `version_introduced`, `enabled`)
VALUES
(700, 'PLG_TWOFACTORAUTH_TRUST_POSTINSTALL_TITLE', 'PLG_TWOFACTORAUTH_TRUST_POSTINSTALL_BODY', '', 'plg_twofactorauth_trust', 1, 'message', '', '', '', '', '3.9.0', 1);
```

Just make sure you replace #_ with your database prefix

After that:

Enable the Two Factor Authentication - Google Authenticator plugin
Enable the Two Factor Authentication - Trust this device plugin
Configure your Google authentication key in the user administrator page or in frontend account
Logout and try to login

**The problems I found**

1) First thing I noticed is the tooltip is trust this device but it is actualy trust this browser because it is based on the browser cookies you are on. So maybe change the language
tooltip string to make it clear it is specific to the browser your are on

trust-tooltip

2) In the plugin page you have this
"Allow user to mark the computer or mobile device as trusted. With trusted computers and devices, you don’t need to enter a verification code each time you sign in."
I think this might confuse users because your're not marking the computer or mobile but the browser as trusted

trust-plugin-page


3) The trust this device checkbox is not available in Hathor admin template login page

4) The checkbok is not centered in the admin login page (Isis template)

trust-device

**Successful tests**

Tested in Chrome, logout and open the admin in Firefox and it does not work unless you check the trust in Firefox too.

Tested cleaning cookies and the trust does not work anymore as expected

Tested with superuser and administrator accounts and both works in the backend when you do the check

Tested disabling the extension in Firefox and tried to login in Chrome and it correctly prompted me to use the two factor secret key

If you check the trust this device in the backend, logout and go to frontend and try to login you can't
you have to check the trust in the frontend too. I think this is expected.

Tested with registered user in frontend and the trust this device works too

Tested the backend in Chrome mobile and it works too

Tested choosing 1 day for the cookie life time and checked 24h later and it showed the warning "The two factor authentication Secret Key is invalid."

------------------------------
I only tested with Google two factor authenticator and with Chrome, FF, Edge and IE 11 latest versions.
I only tested the functionality, not the code as I am not a developer.

I'll mark my test successful as soon the problems I found has been answered.

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on e2446dc70bfc4759edb52d513add0892d9a7b93b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20191">issues.joomla.org/tracker/joomla-cms/20191</a>.</sub>

**Comment by jjnxpct** [comment]
I have tested this item :white_check_mark: successfully on e2446dc70bfc4759edb52d513add0892d9a7b93b<br /><br />Hi! I tested this on/with:

Joomla! 3.9.1-dev 
PHP 7.2.11 | MySQL 5.6.41
Chrome & Firefox
Authentication method: Yubikey

In used the Joomla Patch tester to install the patch on my test CMS. I did need to run the SQL code provided by fabpacheco. But to I did also need to change the SQL query, because the '495' extension ID is now used bu a privacy plugin. So I changed that to '496'.

After this I could activate the plugin (only became visible in the backend after the SQL query) and the tests were successful. A cookie was placed on my computer by the browser when I checked the trust checkbox. The next time I did not need to enter the key again, I could just leave it untouched. 

I would like to suggest, when the cookie is placed, dot not show the key field at all? Or some other way to communicate to the user that a key is not needed at that time?

I did not test different settings for the plugin 'Two Factor Authentication - Trust this device '. I did see the cookie expiration date was set OK (30 days) so I guess that will work fine.

Hope to see this implemented soon.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20191">issues.joomla.org/tracker/joomla-cms/20191</a>.</sub>

**Comment by jjnxpct** [comment]
login_2fa_trust-8

Maybe something like this?

**Comment by B3nito** [comment]
Can the Secret code be put on a second site like real 2 FA is like?

**Comment by mbabker** [comment]
Joomla core doesn't have a "captive" login system that would let you have a secondary login screen like many other sites with a 2FA workflow do (BTW there is nothing "wrong" about having the 2FA field alongside the main login form other than at worst introducing some user confusion about what a "secret key" is).

**Comment by jjnxpct** [comment]
@mbabker That's why I put '2FA key' as a placeholder text in the 2FA field. That might be a bit more explanatory. Also I think the star icon could be replaced with a key icon to better fit the purpose of this field.

**Comment by jjnxpct** [comment]
Is there anything I can do (except actual coding...) to help this issue to get merged into the CMS? Does it need more tests from others? Not sure how this works... I see 'some checks were not succesful'. Are these automated test? Or is someone responsible for these tests? Sorry about these 'newbie' questions.

**Comment by zero-24** [comment]
Nothing to be sorry about. I have just fixed the conflicts and updated to staging (that reboots the tests). So yes please test the code / Plugin and let us know 👍

File: plugins/twofactorauth/trust/trust.xml [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+					label="PLG_TWOFACTORAUTH_TRUST_FIELD_KEY_LENGTH_LABEL"
+					description="PLG_TWOFACTORAUTH_TRUST_FIELD_KEY_LENGTH_DESC"
+					default="16"
+					filter="integer"
+					required="true"
+					>
+					<option value="8">8</option>
+					<option value="16">16</option>
+					<option value="32">32</option>
+					<option value="64">64</option>
```
Comment (bembelimen) [inline plugins/twofactorauth/trust/trust.xml:44]: What's the reason for chosing something below 64? Or could we just set 64 without the select?

**Comment by Quy** [comment]
Rebase for J4?

**Comment by ghost** [comment]
As new Feature it has to be rebased on J4, so closing. Thanks for hint @Quy
